core/src/main/java/kafka/log/remote/RemoteLogManager.java

```diff
@@ -1265,6 +1248,26 @@ void cleanupExpiredRemoteLogSegments() throws RemoteStorageException, ExecutionE
                         canProcess = false;
                         continue;
                     }
+                    // This works as retry mechanism for dangling remote segments.
+                    // A dangling remote segment is one that failed the copy to remote storage and the following cleanup.
+                    // Rather than waiting for the retention to kick in, we cleanup early to avoid polluting the cache and possibly waste remote storage.
+                    if (RemoteLogSegmentState.DELETE_SEGMENT_STARTED.equals(metadata.state())) {
+                        LOGGER.info("Cleaning dangling segment {}", metadata.remoteLogSegmentId());
+                        try {
+                            deleteRemoteLogSegment(
+                                metadata,
+                                ignored -> !isCancelled(),
+                                remoteLogMetadataManager,
+                                remoteLogStorageManager,
+                                brokerTopicStats,
+                                time,
+                                brokerId
+                            );
+                            LOGGER.info("Successfully cleaned dangling segment {}", metadata.remoteLogSegmentId());
+                        } catch (RemoteStorageException e1) {
+                            LOGGER.warn("Error while cleaning dangling segment {}: {}", metadata.remoteLogSegmentId(), e1.getMessage());
+                        }
+                    }
```

Comment:
1. We miss the `continue` in this if block, so it will also enter `isSegmentBreachedByRetentionTime`/ `isSegmentBreachedByRetentionSize` check below and have duplicate deletion.
2. With this change, the `DELETE_SEGMENT_STARTED` segments won't be counted in `RemoteDeleteLag` metrics in L1314 below. I think we should let the purpose of the while loop clear that we're just collecting segments to be deleted, and then, we do all the deletion in L1316 below. 

So I'm thinking we can just do something like this:
```
if (RemoteLogSegmentState.DELETE_SEGMENT_STARTED.equals(metadata.state())) {
  segmentsToDelete.add(metadata);
  continue;
}
```
WDYT?

Reply:
Also, I just checked that we don't have to do the log start offset for `DELETE_SEGMENT_STARTED` segments (i.e. `remoteLogRetentionHandler.logStartOffset.ifPresent(offset -> handleLogStartOffsetUpdate(topicIdPartition.topicPartition(), offset));
`) because:
1. If it is expired segments, but failed to delete the segment, the log start offset should have updated in previous iteration
2. If it's failed copy segment, and failed to delete the segment, in this case, the log start offset should not get updated.

So it should be fine. Does that make sense?

Reply:
Yes, I agree, we can just collect and delete in once place. I also tried this approach but was unsure about the consequences. Thanks.